packages/block-library/src/site-logo/index.php

```diff
@@ -160,3 +160,22 @@ function _delete_site_logo_on_remove_custom_logo_on_setup_theme() {
 	add_action( "delete_option_theme_mods_$theme", '_delete_site_logo_on_remove_theme_mods' );
 }
 add_action( 'setup_theme', '_delete_site_logo_on_remove_custom_logo_on_setup_theme', 11 );
+
+/**
+ * Removes the custom_logo theme-mod when the site_logo option gets deleted.
+ */
+function _delete_custom_logo_on_remove_site_logo() {
+	$theme = get_option( 'stylesheet' );
+
+	// Unhook update and delete actions for custom_logo to prevent a loop of hooks.
```

Comment:
Good idea adding this to avoid accidentally causing a loop of hooks.

✅ Tested that with this code in the Gutenberg plugin, by execution time `has_action` returns `false` so these actions aren't registered, however the code doesn't appear to throw any warnings or cause any issues
✅ Tested that with this code copied to the `site-title.php` file in `public_html/wp-includes/blocks/site-logo.php` the actions _are_ registered, so this code prevents the loop

For testing I added the following and inspected by `debug.log` file:

```
	$has_action_1 = has_action( "update_option_theme_mods_$theme", '_delete_site_logo_on_remove_custom_logo' );
	$has_action_2 = has_action( "delete_option_theme_mods_$theme", '_delete_site_logo_on_remove_theme_mods' );

	error_log( 'has actions' );
	error_log( var_export( $has_action_1, true ) );
	error_log( var_export( $has_action_2, true ) );
```

The results were:

In Gutenberg plugin:

```
[17-Sep-2021 01:22:46 UTC] has actions
[17-Sep-2021 01:22:46 UTC] false
[17-Sep-2021 01:22:46 UTC] false
```

In Core:

```
[17-Sep-2021 01:36:55 UTC] has actions in core!
[17-Sep-2021 01:36:55 UTC] 10
[17-Sep-2021 01:36:55 UTC] 10
```

Reply:
Gutenberg's build process is getting in the way here: because `_delete_site_logo_on_remove_custom_logo` and `_delete_site_logo_on_remove_theme_mods` are function names in this file, [they are being automatically prefixed](https://github.com/WordPress/gutenberg/blob/add/remove-site-logo-from-block/tools/webpack/blocks.js#L108) with `gutenberg_`.

This means that the unprefixed Core hooks for the site logo block are not removed, and end up running with `remove_theme_mod( 'custom_logo' )`, the result being that `delete_option( 'site_logo' );` can be called twice. There's [a check](https://github.com/WordPress/wordpress-develop/blob/0f0089391b0f92d4508b8a4df567df2052161340/src/wp-includes/option.php#L738) that prevents this from becoming an infinite loop, but I still think we should prevent this from happening.

You can verify this with the following steps

- Delete any existing site_logo (block) and custom_logo (Customizer) settings
- Add a custom logo by setting the logo in the Customizer
- Insert the Site Logo block in a post/page (you'll see the custom logo you set)
- Reset the logo using the button added in this PR
- If you add a log call or breakpoint to `_delete_site_logo_on_remove_custom_logo` in https://github.com/WordPress/wordpress-develop/blob/0f0089391b0f92d4508b8a4df567df2052161340/src/wp-includes/blocks/site-logo.php#L135, you'll see this function is still called on the `"update_option_theme_mods_$theme"` hook.

Without changing the build process, we could make the relevant function names in this file different from core, and that would work around the issue... something like

```php
function gutenberg_delete_site_logo_on_remove_custom_logo( $old_value, $value ) { ...

function gutenberg_delete_site_logo_on_remove_theme_mods() { ...

function _delete_custom_logo_on_remove_site_logo() {
	$theme = get_option( 'stylesheet' );

	// Unhook update and delete actions for custom_logo to prevent a loop of hooks.
	// Gutenberg hooks.
	remove_action( "update_option_theme_mods_$theme", 'gutenberg_delete_site_logo_on_remove_custom_logo', 10 );
	remove_action( "delete_option_theme_mods_$theme", 'gutenberg_delete_site_logo_on_remove_theme_mods' );

	// Core hooks.
	remove_action( "update_option_theme_mods_$theme", '_delete_site_logo_on_remove_custom_logo', 10 );
	remove_action( "delete_option_theme_mods_$theme", '_delete_site_logo_on_remove_theme_mods' );

	// Remove the custom logo.
	remove_theme_mod( 'custom_logo' );

	// Restore update and delete actions.
	// Gutenberg hooks.
	add_action( "update_option_theme_mods_$theme", 'gutenberg_delete_site_logo_on_remove_custom_logo', 10, 2 );
	add_action( "delete_option_theme_mods_$theme", 'gutenberg_delete_site_logo_on_remove_theme_mods' );

	// Core hooks.
	add_action( "update_option_theme_mods_$theme", '_delete_site_logo_on_remove_custom_logo', 10, 2 );
	add_action( "delete_option_theme_mods_$theme", '_delete_site_logo_on_remove_theme_mods' );
}
add_action( 'delete_option_site_logo', '_delete_custom_logo_on_remove_site_logo' );
```

It's a little strange, as those functions will become prefixed with `gutenberg_gutenberg_`, but in practice I don't think that matters.

---

Another option would be to remove `_delete_site_logo_on_remove_custom_logo_on_setup_theme` from this file, and the two functions it hooks in (`_delete_site_logo_on_remove_custom_logo` and `_delete_site_logo_on_remove_theme_mods`).

Because this file is included on the `init` hook, and `_delete_site_logo_on_remove_custom_logo_on_setup_theme` is hooked into `setup_theme`, which runs before `init`, these functions are never run from Gutenberg ([only from Core](https://github.com/WordPress/wordpress-develop/blob/0f0089391b0f92d4508b8a4df567df2052161340/src/wp-includes/blocks/site-logo.php#L162)). They seem to be here only to keep this file consistent with Core. (related issue: https://github.com/WordPress/gutenberg/issues/33177)

Reply:
> There's a check that prevents this from becoming an infinite loop, but I still think we should prevent this from happening.

That explains how I missed it 😱 Thank you for checking this!

I went with the first suggestion to rename the hooks in Gutenberg and unhook/rehook both. I'm a bit torn since I agree it's a little strange, but I think doing so actually makes the hook duplication clearer.

Reply:
That fixed the issue, thanks!